### PR TITLE
DDF-2320 Updated CatalogFramework to try all InputTransformers before failing ingest

### PR DESCRIPTION
**Pull request by bdeining**
#### What does this PR do?
When multiple InputTransformers are present that all handle a given mime-type (e.g. text/xml) they should each be called until one successfully creates a metacard. Currently, if any of them throw an exception, the processing chain is stopped and the other InputTransformers are not called.
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@rzwiefel @glenhein 
#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
@jrnorth 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith 
@pklinef
#### How should this be tested? (List steps with links to updated documentation)
Build / Install / log:set DEBUG ddf.catalog.impl / Ingest XML and observe multiple transformers are called / use PostMan to test the rest endpoint
#### What are the relevant tickets?
[DDF-2320](https://codice.atlassian.net/browse/DDF-2320)
#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

**Comment by glenhein** [comment]
How have you tested this? Did you force a transformer to fail to confirm that the other transformers were called?

**Comment by bdeining** [comment]
Yes, you can test by ingesting a `pom.xml` which causes the XmlInputTransformer to fail. 

```
How should this be tested? (List steps with links to updated documentation)

Build / Install / log:set DEBUG ddf.catalog.impl / Ingest XML and observe multiple transformers are called```

**Comment by pklinef** [comment]
@bdeining Can you provide an example of the message for the `MetacardCreationException` that is created by these changes when trying to ingest a pom.xml into DDF?

**Comment by bdeining** [comment]
@pklinef Here's an example of failed transformations at the DEBUG level :
```
08:20:15,835 | DEBUG | qtp305189561-172 | .catalog.impl.operations.MetacardFactory   86 | re-standardframework | Transformer [ddf.catalog.transformer.xml.XmlInputTransformer@4b62fc28] could not create metacard.
ddf.catalog.transform.CatalogTransformerException: Failed Transformation.  Could not create Metacard from XML.
08:20:16,108 | DEBUG | qtp305189561-172 | .catalog.impl.operations.MetacardFactory   86 | re-standardframework | Transformer [org.codice.ddf.spatial.ogc.csw.catalog.transformer.GmdTransformer@5af35c10] could not create metacard.
ddf.catalog.transform.CatalogTransformerException: Unable to transform from GMD Metadata to Metacard
08:20:16,166 | DEBUG | qtp305189561-172 | .catalog.impl.operations.MetacardFactory   86 | re-standardframework | Transformer [org.codice.ddf.spatial.ogc.csw.catalog.converter.CswRecordConverter@7a84a78f] could not create metacard.
ddf.catalog.transform.CatalogTransformerException: Unable to transform from CSW Record to Metacard.
```

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java [hunk 1]
```diff
@@ -59,50 +63,39 @@ Metacard generateMetacard(String mimeTypeRaw, String id, String fileName, Subjec
...
             }
-            LOGGER.debug("Transformer [{}] could not create metacard.", transformer, e);
         }
 
         if (generatedMetacard == null) {
-            throw new MetacardCreationException(causeMessage.toString());
+            throw new MetacardCreationException(String.format(
+                    "Could not create metacard with mimeType %s : %s",
+                    mimeTypeRaw,
+                    stackTraceList.stream()
```
Comment (jrnorth) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java]: You can also do `String.join("\n", stackTraceList)`

**Comment by jlcsmith** [comment]
@bdeining  - I'll let you respond to @jrnorth  before I merge.

**Comment by brendan-hofmann** [comment]
I think `RESTEndpoint` and `IngestCommand` have this same issue. Do you want to fix those as well, or refactor them to use the correct code in `CatalogFramework` instead?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java (hunk 1 above)
Comment (bdeining) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java]: Changed

**Comment by bdeining** [comment]
@brendan-hofmann.  This PR fixes the issue on the `RESTEndpoint`.  Since this PR is wanted for the 2.10.0 release, the `IngestCommand` should be refactored under a different ticket : https://codice.atlassian.net/browse/DDF-2714

**Comment by brendan-hofmann** [comment]
@bdeining I was thinking of `RESTEndpoint.generateMetacard`, which also iterates through the available transformers. Re-reading it, it looks like it does try all available transformers before failing though. I'm good with fixing the `IngestCommand` later since it's non-critical and has a workaround.

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java [hunk 2]
```diff
@@ -56,50 +59,38 @@ Metacard generateMetacard(String mimeTypeRaw, String id, String fileName, Path t
...
+        MimeType mimeType = new MimeType(mimeTypeRaw);
+
+        List<InputTransformer> listOfCandidates = mimeTypeToTransformerMapper.findMatches(
+                InputTransformer.class,
+                mimeType);
+        List<String> stackTraceList = new ArrayList<>();
+
+        LOGGER.debug("List of matches for mimeType [{}]: {}", mimeType, listOfCandidates);
+
+        for (InputTransformer candidate : listOfCandidates) {
```
Comment (brendan-hofmann) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java]: This seems to make the logic in https://github.com/codice/ddf/blob/5113a905dab89696fe97f18418424397d5e50ace/catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java#L1161-L1161 redundant, or at the very least it's code duplication. Can we remove the logic from the `RESTEndpoint.generateMetacard`, or dedupe the code if that's not feasible?

Comment (bdeining) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java]: After looking at this, it's not easily feasible to remove this logic (the MetacardFactory methods are protected, not published as a service, etc).  I've refactored the code to match the code changes in the MetacardFactory.

**Comment by brendan-hofmann** [comment]
Successfully heroed. :shipit:

**Pull request merged**